File: aat/cpp/src/core/orderbook/order_book.cpp
#include <aat/core/order_book/order_book.hpp>

namespace aat {
namespace core {
    
    OrderBookIterator&
    OrderBookIterator::operator++() {
        // TODO

        return *this;
    }

    std::shared_ptr<Order>
    OrderBookIterator::operator*() {
        if (side == Side::SELL) {
            return (*(order_book.sells.at(price_level)))[index_in_level];
        }
        else {
            return (*(order_book.buys.at(price_level)))[index_in_level];
        }
    }

    bool 
    OrderBookIterator::operator==(const OrderBookIterator &that) {
        //TODO
        return false;
    }

    OrderBook::OrderBook(const Instrument &instrument)
        : instrument(instrument)
        , exchange(NullExchange)
        , callback(nullptr) {}

    OrderBook::OrderBook(const Instrument &instrument, const ExchangeType &exchange)
        : instrument(instrument)
        , exchange(exchange)
        , callback(nullptr) {}

    OrderBook::OrderBook(
        const Instrument& instrument, const ExchangeType& exchange, std::function<void(std::shared_ptr<Event>)> callback)
        : instrument(instrument)
        , exchange(exchange)
        , callback(callback) {}

    void 
    OrderBook::setCallback(std::function<void(std::shared_ptr<Event>)> callback){
        collector.setCallback(callback);
    }

    void 
    OrderBook::reset() {
        buy_levels = std::vector<double>();
        sell_levels = std::vector<double>();
        buys = std::unordered_map<double, std::shared_ptr<PriceLevel>>();
        sells = std::unordered_map<double, std::shared_ptr<PriceLevel>>();
        collector = Collector(callback);
    }

    void 
    OrderBook::add(std::shared_ptr<Order> order) {
        // secondary triggered orders
        std::vector<std::shared_ptr<Order>> secondaries;
        
        // order is buy, look at top of sell side
        double top = getTop(order->side, collector.getClearedLevels());
        
        //set levels to the right side
        std::vector<double> &levels = (order->side == Side::BUY) ? buy_levels : sell_levels;
        std::unordered_map<double, std::shared_ptr<PriceLevel>> &prices = (order->side == Side::BUY) ? buys : sells;
        std::unordered_map<double, std::shared_ptr<PriceLevel>> &prices_cross = (order->side == Side::BUY) ? sells : buys;

        // set order price appropriately
        double order_price;
        if (order->order_type == OrderType::MARKET) {
            if (order->flag == OrderFlag::NONE) {
                // price goes infinite "fill however you want"
                order_price = 
                    (order->side == Side::BUY) ? std::numeric_limits<double>::max() : std::numeric_limits<double>::min();
            }
            else {
                // with a flag, it will have a price
                order_price = order->price;
            }
        }
        else {
            order_price = order->price;
        }
        
        // Check if crosses
        while (top > 0.0 && ((order->side == Side::BUY) ? order_price >= top : order_price <= top)) {
            // execute order against level
            // if returns trade, it cleared the level
            // else, order was fully executed
            std::shared_ptr<Order> trade = prices_cross[top]->cross(order, secondaries);
            if (trade) {
                // clear sell level
                top = getTop(order->side, collector.clearLevel(prices_cross[top]));
                continue;
            }

            // trade is done, check if level was cleared exactly
            if (prices_cross.at(top)->size() <= 0) {
                //level cleared exactly
                collector.clearLevel(prices_cross.at(top));
            }
            break;
        }

        // if order remaining, check rules / push to book
        if (order->filled < order->volume) {
            if (order->order_type == OrderType::MARKET) {
                // market orders
                if (order->flag == OrderFlag::ALL_OR_NONE || order->flag == OrderFlag::FILL_OR_KILL) {
                    // cancel the order, do not execute any
                    collector.revert();
                    
                    // cancel the order
                    collector.pushCancel(order);
                    collector.commit();
                } else {
                    // market order, partial
                    if (order->filled > 0)
                        collector.pushTrade(order, order->filled);
                    
                    // clear levels
                    clearOrders(order, collector.getClearedLevels());
                    
                    // execute order
                    collector.pushCancel(order);
                    collector.commit();
                    
                    // execute secondaries
                    for (std::shared_ptr<Order> secondary : secondaries) {
                        secondary->timestamp = order->timestamp;
                        add(secondary);
                    }
                }
            } else {
                // limit orders
                if (order->flag == OrderFlag::FILL_OR_KILL) {
                    if (order->filled > 0) {
                        // reverse partial, canncel the order, do not execute any
                        collector.revert();
                        //cancel the order
                        collector.pushCancel(order);
                        collector.commit();
                    } else {
                        // add to book
                        collector.commit();

                        // limit order, put on books
                        if (insort(levels, order->price)) {
                            // new price level
                            prices[order->price] = std::make_shared<PriceLevel>(order->price, collector);
                        }
                        
                        // add order to price level
                        prices[order->price]->add(order);

                        // execute secondaries
                        for (std::shared_ptr<Order> secondary : secondaries) {
                            secondary->timestamp = order->timestamp;
                            add(secondary);
                        }
                    }
                } else if (order->flag == OrderFlag::ALL_OR_NONE) {
                    if (order->filled > 0) {
                        // order could not fill fully, revert
                        // cancel the order, do not execute any
                        collector.revert();
                        collector.pushCancel(order);
                        collector.commit();
                    } else {
                        // add to book
                        collector.commit();

                        // limit order, put on books
                        if (insort(levels, order->price)) {
                            // new price level
                            prices[order->price] = std::make_shared<PriceLevel>(order->price, collector);
                        }
                        
                        // add order to price level
                        prices[order->price]->add(order);

                        // execute secondaries
                        for (std::shared_ptr<Order> secondary : secondaries) {
                            secondary->timestamp = order->timestamp;
                            add(secondary);
                        }
                    }
                } else if (order->flag == OrderFlag::IMMEDIATE_OR_CANCEL) {
                    if (order->filled > 0) {
                        // clear levels
                        clearOrders(order, collector.getClearedLevels());

                        // execute the ones that filled, kill the remainder
                        collector.pushCancel(order);
                        collector.commit();

                        // execute secondaries
                        for (std::shared_ptr<Order> secondary : secondaries) {
                        secondary->timestamp = order->timestamp;  // adjust trigger time
                        add(secondary);
                        }

                    } else {
                        // add to book
                        collector.commit();

                        // limit order, put on books
                        if (insort(levels, order->price)) {
                        // new price level
                        prices[order->price] = std::make_shared<PriceLevel>(order->price, collector);
                        }
                        // add order to price level
                        prices[order->price]->add(order);

                        // execute secondaries
                        for (std::shared_ptr<Order> secondary : secondaries) {
                        secondary->timestamp = order->timestamp;  // adjust trigger time
                        add(secondary);
                        }
                    }
                } else {
                    // clear levels
                    clearOrders(order, collector.getClearedLevels());

                    // execute order
                    collector.commit();

                    // limit order, put on books
                    if (insort(levels, order->price)) {
                        // new price level
                        prices[order->price] = std::make_shared<PriceLevel>(order->price, collector);
                    }

                    // add order to price level
                    prices[order->price]->add(order);

                    // execute secondaries
                    for (std::shared_ptr<Order> secondary : secondaries) {
                        secondary->timestamp = order->timestamp;  // adjust trigger time
                        add(secondary);
                    }
                }
            }
        } else {
            // don't need to add trade as this is done in the price_levels
            
            // clear levels
            clearOrders(order, collector.getClearedLevels());
            
            // execute all orders
            collector.commit();

            // execute secondaries
            for (std::shared_ptr<Order> secondary : secondaries) {
                secondary->timestamp = order->timestamp;
                add(secondary);
            }
        }
        
        // clear the collector
        collector.clear();
    }





    

    double
    OrderBook::getTop(Side side, uint_t cleared) {
        if (side == Side::BUY) {
            if (sell_levels.size() > cleared) {
                return sell_levels[cleared];
            }
            else {
                return -1;
            }
        }
        else {
            if (buy_levels.size() > cleared) {
                return buy_levels[buy_levels.size()-cleared-1];
            }
            else {
                return -1;
            }
        }
    }
    
} // namespace core
} // namespace aat

File: aat/core/order_book/order_book/order_book.py
from queue import Queue
from typing import (
    Any,
    Callable,
    cast,
    Iterator,
    List,
    Dict,
    Optional,
    Tuple,
    Type,
    Union,
)

from aat.core import ExchangeType, Order, Instrument, Event
from aat.config import Side, OrderFlag, OrderType
from aat.core.data import Order

from ..base import OrderBookBase
from ..cpp import _CPP, _make_cpp_orderbook
from ..collector import _Collector
from ..price_level import _PriceLevel, PriceLevelRO
from ..utils import _insort


class OrderBook(OrderBookBase):
    """
    Args:
        instrument: The instrument of the book
        exchange_name: The name of the exchange
        callback(Function): callback on events
    """

    def __new__(cls: Type, *args: Any, **kwargs: Any) -> "OrderBook":
        if _CPP:
            return _make_cpp_orderbook(*args, **kwargs)
        return super(OrderBook, cls).__new__(cls)
    
    def __init__(self,
                 instrument: Instrument,
                 exchange_name: Union[ExchangeType, str] = "",
                 callback: Optional[callable] = None) -> None:
        self._instrument = instrument
        self._exchange_name = (
            exchange_name if isinstance(exchange_name, ExchangeType)
            else ExchangeType(exchange_name or "")
        )
        self._callback = callback

        # reset levels and collectors
        self.reset()
        
        # default callback is to enqueue
        self._queue: Queue[Event] = Queue()

    @property
    def instrument(self) -> Instrument:
        return self._instrument

    @property
    def exchange(self) -> ExchangeType:
        return self._exchange_name

    @property
    def callback(self) -> Callable:
        return self._callback

    @property
    def queue(self) -> Queue:
        return self._queue
    
    def _push(self, event: Event) -> None:
        self._queue.put(event)

    def reset(self) -> None:
        """
        Reset the orderbook to its base state
        """
        # levels look like [10, 10.5, 11, 11.5]
        self._buy_levels: List[float] = []
        self._sell_levels: List[float] = []

        # look like {price level: PriceLevel}
        self._buys: Dict[float, _PriceLevel] = {}
        self._sells: Dict[float, _PriceLevel] = {}
        
        # setup collector for conditional orders
        self._collector = _Collector(self._callback)

    def setCallback(self, callback: Callable) -> None:
        self._callback = callback
        self._collector.setCallback(callback)

    def find(self, order: Order) -> Optional[Order]:
        """find an order in the order book
        Args:
            order (Data): order to find in orderbook
        """

        price = order.price
        side = order.side
        levels = self._buy_levels if side == Side.BUY else self._sell_levels
        prices = self._buys if side == Side.BUY else self._sells

        if price not in levels:
            return None
        
        # find order from pricelevel
        return prices[price].find(order)
    
    def topOfBook(self) -> Dict[Side, PriceLevelRO]:
        """return top of both sides

        Args:

        Returns:
            value (dict): returns {BUY: tuple, SELL: tuple}
        """

        return {
            Side.BUY: cast(PriceLevelRO, self.bids(levels=0)),
            Side.SELL: cast(PriceLevelRO, self.asks(levels=0)),
        }

    def spread(self) -> float:
        """return the spread

        Args:

        Returns:
            value (float): spread between bid and ask
        """
        tob: Dict[Side, PriceLevelRO] = self.topOfBook()
        return tob[Side.BUY] - tob[Side.SELL]
    
    def level(self, level: int = 0, price: Optional[float] = None) -> Tuple:
        """return book level

        Args:
            level (int): depth of book to return
            price (float): price level to look for
        Returns:
            value (tuple): returns ask, bid
        """
        # collect bids and asks at `level`
        if price:
            return (
                PriceLevelRO(
                    self._sells[price].price,
                    self._sells[price].volume,
                    len(self._sells[price])
                )
                if price in self._sell_levels
                else None,
                PriceLevelRO(
                    self._buys[price].price,
                    self._buys[price].volume,
                    len(self._buys[price])
                )
                if price in self._buy_levels
                else None
            )
        
        return (
            PriceLevelRO(
                self._sell_levels[level],
                self._sells[self._sell_levels[level]].volume,
                len(self._sells[self._sell_levels[level]]),
                self._sells[self._sell_levels[level]]._orders
            )
            if len(self._sell_levels) > level
            else PriceLevelRO(0.0, 0.0, 0),
            PriceLevelRO(
                self._buy_levels[-level - 1],
                self._buys[self._buy_levels[-level - 1]].volume,
                len(self._buys[self._buy_levels[-level - 1]]),
                self._buys[self._buy_levels[-level - 1]]._orders,
            )
            if len(self._buy_levels) > level
            else PriceLevelRO(0.0, 0.0, 0),
        )
            
        
    def bids(
        self, levels: int = 0
    ) -> Union[PriceLevelRO, List[Optional[PriceLevelRO]]]:
        """return bid levels starting at top

        Args:
            levels (int): number of levels to return
        Returns:
            value (dict of list): returns [levels in order] for `levels` number of levels
        """
        if levels <= 0:
            return (
                PriceLevelRO(
                    self._buy_levels[-1],
                    self._buys[self._buy_levels[-1]].volume,
                    len(self._buys[self._buy_levels[-1]]),
                    self._buys[self._buy_levels[-1]]._orders,
                )
                if len(self._buy_levels) > 0
                else PriceLevelRO(0, 0, 0)
            )
        return [
            PriceLevelRO(
                self._buy_levels[-i - 1],
                self._buys[self._buy_levels[-i - 1]].volume,
                len(self._buys[self._buy_levels[-i - 1]]),
                self._buys[self._buy_levels[-i - 1]]._orders,
            )
            if len(self._buy_levels) > i
            else None
            for i in range(levels)
        ]

    def asks(
        self, levels: int = 0
    ) -> Union[PriceLevelRO, List[Optional[PriceLevelRO]]]:
        """return ask levels starting at top

        Args:
            levels (int): number of levels to return
        Returns:
            value (dict of list): returns [levels in order] for `levels` number of levels
        """
        if levels <= 0:
            return (
                PriceLevelRO(
                    self._sell_levels[0],
                    self._sells[self._sell_levels[0]].volume,
                    len(self._sells[self._sell_levels[0]]),
                    self._sells[self._sell_levels[0]]._orders,
                )
                if len(self._sell_levels) > 0
                else PriceLevelRO(float("inf"), 0, 0)
            )
        return [
            PriceLevelRO(
                self._sell_levels[i],
                self._sells[self._sell_levels[i]].volume,
                len(self._sells[self._sell_levels[i]]),
                self._sells[self._sell_levels[i]]._orders,
            )
            if len(self._sell_levels) > i
            else None
            for i in range(levels)
        ]
    
    def levels():
        pass

    def cancel():
        pass

    def _clearOrders():
        pass

    def _getTop():
        pass

    


        